feat(replay): add viewed-by endpoint + abstract kafka utils for replay events

aliu39 opened the pull request:
Relates to https://github.com/getsentry/sentry/issues/64924

[Blueprint](https://github.com/getsentry/sentry/blob/bd8b0486473cf3136ba04169028ed76313af51ce/src/sentry/replays/blueprints/api.md?plain=1#L615)

aliu39 committed:
feat(replay): add viewed-by endpoint

iamrajjoshi commented on src/sentry/replays/endpoints/organization_replay_viewed_by.py:
```diff
@@ -0,0 +1,89 @@
+        #     return Response({"data": response[0]}, status=200)
+
+    # TODO:
+    # def post(self, request: Request, organization: Organization):
```
Can we remove the commented code?

aliu3ntry replied:
this is a draft, will mark it 😢

aliu39 committed:
Merge branch 'master' into aliu/viewed-replay-endpoint

aliu39 committed:
Merge branch 'master' of github.com:getsentry/sentry into aliu/viewed-replay-endpoint

aliu39 committed:
draft has_viewed query

aliu39 committed:
rename to has_viewed

aliu39 committed:
add viewed-by to urls

aliu39 committed:
Merge branch 'master' into aliu/viewed-replay-endpoint

aliu39 committed:
add has_viewed to replay details responses

aliu39 committed:
add publish_replay_viewed

aliu39 committed:
viewed-by: post and get drafts

aliu39 committed:
update GET with serializer (2 potential approaches)

aliu39 committed:
rename to actor

aliu39 committed:
update POST with project id query

JoshFerge commented on src/sentry/replays/endpoints/organization_replay_viewed_by.py:
```diff
@@ -0,0 +1,132 @@
+        if not filter_params["start"] or not filter_params["end"]:
+            return Response(status=404)
+
+        # TODO: are there more efficient ways to lookup or pass in the project_id, rather than a snuba query?
```
yeah, why don't we just pass in the project_id? could even make this a project scoped endpoint

aliu3ntry replied:
So move both GET and POST to a project endpoint? Makes sense to me. I was following the blueprint for this draft

JoshFerge commented on src/sentry/replays/tasks.py:
```diff
@@ -104,3 +104,36 @@ def archive_replay(publisher: KafkaPublisher, project_id: int, replay_id: str) -
+    max_retries=5,
+    silo_mode=SiloMode.REGION,
+)
+def publish_replay_viewed(project_id: int, replay_id: str, viewed_by_id: int) -> None:
```
do we want to put some kind of rate limiting on this endpoint per user? what if a user views a replay 100000 times with a selenium script or something?

aliu3ntry replied:
Mm sounds smart. How do we implement this?

JoshFerge commented on src/sentry/replays/post_process.py:
```diff
@@ -73,13 +81,27 @@ class ReplayDetailsResponse(TypedDict, total=False):
-    response: list[dict[str, Any]], fields: list[str]
+    response: list[dict[str, Any]],
+    fields: list[str],
+    actor_uid: int | None = None,
```
why are we calling this actor_uid inster of just user_id or something?

aliu3ntry replied:
Was worried bout confusion with the user_id field in the replays table. Open to suggestions. I also considered:
- as_user
- request_user_id
- viewed_by_id

JoshFerge commented on src/sentry/replays/post_process.py:
```diff
@@ -190,6 +220,32 @@ def generate_sorted_urls(url_groups: list[tuple[int, list[str]]]) -> Iterator[st
+    """Fetch user objects from postgres, serialize them, and format to the output expected by ReplayViewedByEndpoint."""
+    global user_serializer
+
+    # TODO: what happens if a viewed_by_id is invalid or unauthorized?
```
how could it be unauthorized?

aliu3ntry replied:
Didn't get to the bottom of it but I suspected, from the as_user param of serialize_many, that some kind of authorization is needed for some user objects.

aliu39 committed:
Merge branch 'master' of github.com:getsentry/sentry into aliu/viewed-replay-endpoint

aliu39 committed:
change to project endpoint

aliu39 committed:
doc

cmanallen commented on src/sentry/replays/post_process.py:
```diff
@@ -33,7 +41,7 @@ class BrowserResponseType(TypedDict, total=False):
 
 class UserResponseType(TypedDict, total=False):
-    id: str | None
+    id: str | None  # string uint64
```
This id field is not a uint64.  Its a string.  It comes from the customer's application.

cmanallen commented on src/sentry/replays/usecases/query/__init__.py:
```diff
@@ -286,7 +286,12 @@ def make_full_aggregation_query(
 ) -> Query:
-    """Return a query to fetch every replay in the set."""
+    """
+    Return a query to fetch every replay in the set.
```
Remove the line break and start the summary sentence on the first line.  Next to `"""`.

cmanallen commented on src/sentry/replays/endpoints/project_replay_viewed_by.py:
```diff
@@ -0,0 +1,107 @@
+        # query for user ids who viewed the replay
+        filter_params = self.get_filter_params(request, project)
+        # TODO: do we need this range filter?
+        if not filter_params["start"] or not filter_params["end"]:
```
This is already handled by `get_filter_params`.

cmanallen commented on src/sentry/replays/post_process.py:
```diff
@@ -190,6 +220,32 @@ def generate_sorted_urls(url_groups: list[tuple[int, list[str]]]) -> Iterator[st
+
+    # TODO: what happens if a viewed_by_id is invalid or unauthorized?
+
+    # method 1: hybrid cloud user_service. Based on GroupSeenSerializer. No way
```
We should use the hybrid-cloud services.

cmanallen commented on src/sentry/replays/query.py:
```diff
@@ -703,6 +734,11 @@ def _empty_uuids_lambda():
         alias="count_infos",
     ),
+    "viewed_by_ids": Function(
+        "groupUniqArray",
```
We need to remove the `0` ids.  We can ues `groupUniqArrayIf` which accepts a second argument which can be `x -> x > 0`.  Or something similar.  There's prior art in this file with `anyIf`.

aliu3ntry commented on src/sentry/replays/post_process.py:
```diff
@@ -181,6 +203,14 @@ def generate_normalized_output(
+        if actor_uid is not None:
+            if viewed_by_ids := item.pop("viewed_by_ids", None):
+                has_viewed = actor_uid in viewed_by_ids
+        ret_item["has_viewed"] = has_viewed
```
discussed on call

aliu39 committed:
add has_viewed to details and index

aliu39 committed:
Merge branches 'aliu/viewed-replay-endpoint' and 'master' of github.com:getsentry/sentry into aliu/viewed-replay-endpoint

aliu39 committed:
Merge branch 'aliu/replay-details-has-viewed' into aliu/viewed-replay-endpoint

aliu39 committed:
update get

aliu39 committed:
publish replay_viewed synchronously

cmanallen committed:
Merge branch 'master' into aliu/viewed-replay-endpoint

cmanallen committed:
Abstract message publishing into its own module

cmanallen committed:
Add coverage

aliu39 committed:
Merge branch 'master' of github.com:getsentry/sentry into aliu/viewed-replay-endpoint

aliu39 committed:
fix typing and tweak tests

cmanallen committed:
Return structured content

cmanallen committed:
Correct test case

aliu39 committed:
update post test and work on get test

aliu39 committed:
update mock replay viewed

aliu39 committed:
Merge branch 'aliu/viewed-replay-endpoint' of github.com:getsentry/sentry into aliu/viewed-replay-endpoint

aliu39 committed:
fix get test

aliu39 committed:
update tests

aliu39 committed:
add user fields test

aliu39 committed:
tweak response type

aliu39 committed:
rm todos

aliu39 committed:
Merge branch 'master' into aliu/viewed-replay-endpoint

aliu39 committed:
fix publish_replay_event and use in archive_replay

aliu39 committed:
Merge branch 'aliu/viewed-replay-endpoint' of github.com:getsentry/sentry into aliu/viewed-replay-endpoint

aliu39 committed:
rm id field from response and update query_replay_viewed_by_ids

aliu3ntry commented:
Can I get some feedback on the `test_get_replay_viewed_by_user_fields`? Some of these fields are None or empty in the test environment. Wasn't sure how to test all of them, or whether it's necessary to.

aliu39 committed:
docstrings for openapi

aliu3ntry commented:
Also, not sure about the use of @extend_schema_serializer() for ReplayViewedByResponse

evanpurkhiser committed:
fix(crons): Support timezone in MockTimelineVisualization (#68477)

The visualization will now provide the timezone to the sample schedule
data endpoint.

Refs GH-68473

Abdkhan14 committed:
fix(new-trace-url) Updated node path encoding. (#68458)

Trace view page expects `[`span-${spanId}`, `txn-${eventId}`]` to look
for and scroll to span but we were passing [`span:${spanId}`,
`txn:${eventId}`]

Co-authored-by: Abdullah Khan <abdullahkhan@PG9Y57YDXQ.local>

cleptric committed:
Simplify Laravel Metrics onboarding (#68513)

loewenheim committed:
fix(proguard): Filter empty stacktraces in A/B test (#68445)

I've seen a case where during symbolication, the event has 2
stacktrace_infos, but then during the A/B test in the plugin, it has 3.
The plugin counts stacktrace_infos with a `None` stacktrace (which
results from an exception with no stacktrace), but
`find_stacktraces_in_data`, which is used in symbolication, filters
them. This discrepancy results in spurious A/B test failures.

We now filter empty stacktrace_infos in A/B tests as well.

loewenheim committed:
fix(proguard): Correctly merge symbolicated and unsymbolicated exceptions (#68421)

We don't send all exceptions to Symbolicator. This means that obviously
we can't just zip the exceptions in the event and those we get back from
Symbolicator because they don't correspond one-to-one.

ArthurKnaus committed:
fix(area-chart): Revert accidental commit (#68521)

Revert (accidentally committed?) changes to the area chart.

cleptric committed:
Add Laravel & Symfony to metrics opt-in modal (#68522)

cmanallen committed:
Remove assignment

cmanallen committed:
Return empty response without calling

cmanallen committed:
Revert "fix publish_replay_event and use in archive_replay"

This reverts commit f2f146bbdbf4fd158820ef001161d572b3a60e09.

jjbayer committed:
feat(spans): Extraction sample rate (#68527)

See https://github.com/getsentry/relay/pull/3394.

cmanallen committed:
Specify channel

cmanallen committed:
Move user serialization out of post_process and into endpoint

cmanallen committed:
Remove comment

Commit:
ref(ddm): Implement simpler querying structure and improve code (#68436)

Swatinem committed:
Hide "Processing Issues" setting if it was never activated (#68173)

As one step towards deprecating "Legacy Reprocessing", this hides the
"Processing Issues" settings page completely if it has never been active
before.

It is still showing up for projects which have it active, or have ever
had it active.

Along with that, it will now display a deprecation notice, though I
might remove it from this PR if it is too early for that.

shruthilayaj committed:
chore: handle null segment_id since it's optional (#68460)

handle null segment_id since it's optional
stop logging payload in process_spans
some messages take over 250ms to process, add one more span to confirm
its the deserialize

cmanallen committed:
Assert replay_id returned

cmanallen committed:
Add message generator coverage

cmanallen committed:
Match user_service response output to the blueprint

gggritso committed:
feat(perf): Synchronize HTTP sample chart and table (#68454)

Works just like the Queries and Resources panel! In fact, I had to
introduce some temporary duplication to make it work, but in a more
generic form. I'll step back and resolve the duplication in the next
week or so.

In order to fully fix this, I need to make it possible to pass a regular
plain-old `scatter` series to a chart, which is mysteriously impossible
right now, and also doesn't have proper tooltip support.

So many crimes to undo, so little time!

- Add `onHighlight` support to `DurationChart`
- Add simple debounced state hook
- Sync the sample chart and table highlight

aliu3ntry commented on src/sentry/replays/endpoints/project_replay_viewed_by.py:
```diff
@@ -104,8 +110,46 @@ def post(self, request: Request, project: Project, replay_id: str) -> Response:
+    The viewed-by resource is expected to return a subset of the user_service's
+    response output.
+    """
+    return {
```
What are the chances the output of user_service.serialize changes, and this breaks? Should we use .get() instead?

aliu3ntry commented on src/sentry/replays/endpoints/project_replay_viewed_by.py:
```diff
@@ -0,0 +1,155 @@
+
+        viewed_by_ids = viewed_by_ids_response[0]["viewed_by_ids"]
+        if viewed_by_ids == []:
+            return Response({"data": {"id": replay_id, "viewed_by": []}}, status=200)
```
Changed the blueprint yesterday to omit the id

aliu39 committed:
Remove id field, fix response type, and small comment

aliu39 committed:
Update comment

cmanallen committed:
Update docstrings

cmanallen committed:
Fix docs

cmanallen commented:
Sorry for the pings everyone...

cmanallen committed:
Merge branch 'master' into aliu/viewed-replay-endpoint
